File: go/v2/crypto/aead_test.go
package crypto

import (
	"bytes"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

type goldenAEADBasic struct {
	Description string `json:"description"`
	KeyHex      string `json:"key_hex"`
	NonceHex    string `json:"nonce_hex"`
	PlaintextB64 string `json:"plaintext_b64"`
	AADB64       string `json:"aad_b64"`
	ExpectedCiphertextB64 string `json:"expected_ciphertext_b64"`
	ExpectedTagB64        string `json:"expected_tag_b64"`
}

type goldenAEADWrap struct {
	Description string `json:"description"`
	WrapKeyB64  string `json:"wrap_key_b64"`
	WrapNonceB64 string `json:"wrap_nonce_b64"`
	MasterKeyB64 string `json:"master_key_b64"`
	ExpectedWrappedKeyB64 string `json:"expected_wrapped_key_b64"`
}

func TestAEADBasicEncrypt(t *testing.T) {
	data, err := os.ReadFile(filepath.Join("testdata", "golden", "aead", "basic_encrypt.json"))
	if err != nil {
		t.Fatalf("读取 golden 失败: %v", err)
	}
	var g goldenAEADBasic
	if err := json.Unmarshal(data, &g); err != nil {
		t.Fatalf("解析 golden 失败: %v", err)
	}

	key, err := hex.DecodeString(g.KeyHex)
	if err != nil {
		t.Fatalf("解码 key_hex: %v", err)
	}
	nonce, err := hex.DecodeString(g.NonceHex)
	if err != nil {
		t.Fatalf("解码 nonce_hex: %v", err)
	}
	plaintext, err := base64.StdEncoding.DecodeString(g.PlaintextB64)
	if err != nil {
		t.Fatalf("解码 plaintext: %v", err)
	}
	aad, err := base64.StdEncoding.DecodeString(g.AADB64)
	if err != nil {
		t.Fatalf("解码 aad: %v", err)
	}
	expectedCT, err := base64.StdEncoding.DecodeString(g.ExpectedCiphertextB64)
	if err != nil {
		t.Fatalf("解码 expected_ciphertext: %v", err)
	}
	expectedTag, err := base64.StdEncoding.DecodeString(g.ExpectedTagB64)
	if err != nil {
		t.Fatalf("解码 expected_tag: %v", err)
	}

	ct, tag, err := AESGCMEncrypt(key, nonce, plaintext, aad)
	if err != nil {
		t.Fatalf("AESGCMEncrypt: %v", err)
	}
	if !bytes.Equal(ct, expectedCT) {
		t.Fatalf("ciphertext 不匹配\n期望: %x\n实际: %x", expectedCT, ct)
	}
	if !bytes.Equal(tag, expectedTag) {
		t.Fatalf("tag 不匹配\n期望: %x\n实际: %x", expectedTag, tag)
	}

	// 解密回环
	pt, err := AESGCMDecrypt(key, nonce, ct, tag, aad)
	if err != nil {
		t.Fatalf("AESGCMDecrypt: %v", err)
	}
	if !bytes.Equal(pt, plaintext) {
		t.Fatalf("解密结果不匹配\n期望: %x\n实际: %x", plaintext, pt)
	}
}

func TestAEADWrapMasterKey(t *testing.T) {
	data, err := os.ReadFile(filepath.Join("testdata", "golden", "aead", "wrap_master_key.json"))
	if err != nil {
		t.Fatalf("读取 golden 失败: %v", err)
	}
	var g goldenAEADWrap
	if err := json.Unmarshal(data, &g); err != nil {
		t.Fatalf("解析 golden 失败: %v", err)
	}

	wrapKey, _ := base64.StdEncoding.DecodeString(g.WrapKeyB64)
	wrapNonce, _ := base64.StdEncoding.DecodeString(g.WrapNonceB64)
	masterKey, _ := base64.StdEncoding.DecodeString(g.MasterKeyB64)
	expectedWrapped, _ := base64.StdEncoding.DecodeString(g.ExpectedWrappedKeyB64)

	// V2 wrapped_key = ciphertext || tag，aad 为空
	ct, tag, err := AESGCMEncrypt(wrapKey, wrapNonce, masterKey, nil)
	if err != nil {
		t.Fatalf("AESGCMEncrypt: %v", err)
	}
	got := append(append([]byte{}, ct...), tag...)
	if !bytes.Equal(got, expectedWrapped) {
		t.Fatalf("wrapped_key 不匹配\n期望: %x\n实际: %x", expectedWrapped, got)
	}

	// 解密回环
	if len(got) < AESGCMTagLen {
		t.Fatal("wrapped_key 长度异常")
	}
	n := len(got) - AESGCMTagLen
	pt, err := AESGCMDecrypt(wrapKey, wrapNonce, got[:n], got[n:], nil)
	if err != nil {
		t.Fatalf("AESGCMDecrypt: %v", err)
	}
	if !bytes.Equal(pt, masterKey) {
		t.Fatalf("解密 master_key 不匹配")
	}
}

// TestAEADTamper 验证篡改 ciphertext / tag / aad 会触发解密失败。
func TestAEADTamper(t *testing.T) {
	key := bytes.Repeat([]byte{0x11}, 32)
	nonce := bytes.Repeat([]byte{0x22}, 12)
	pt := []byte("AUN V2 message")
	aad := []byte("aad-bytes")

	ct, tag, err := AESGCMEncrypt(key, nonce, pt, aad)
	if err != nil {
		t.Fatalf("encrypt: %v", err)
	}

	// 篡改 ct
	bad := append([]byte{}, ct...)
	bad[0] ^= 0x01
	if _, err := AESGCMDecrypt(key, nonce, bad, tag, aad); err == nil {
		t.Fatal("篡改 ct 应解密失败")
	}
	// 篡改 tag
	badTag := append([]byte{}, tag...)
	badTag[0] ^= 0x01
	if _, err := AESGCMDecrypt(key, nonce, ct, badTag, aad); err == nil {
		t.Fatal("篡改 tag 应解密失败")
	}
	// 篡改 aad
	if _, err := AESGCMDecrypt(key, nonce, ct, tag, []byte("other-aad")); err == nil {
		t.Fatal("篡改 aad 应解密失败")
	}
}

File: js/tests/conformance/aead.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync, readdirSync } from 'fs';
import { join } from 'path';
import { aesGcmEncrypt, aesGcmDecrypt } from '../../src/v2/crypto/aead';

function b64(s: string): Uint8Array {
  const bin = atob(s);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

function hex(s: string): Uint8Array {
  if (s.length % 2 !== 0) throw new Error('hex length must be even');
  const out = new Uint8Array(s.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(s.substr(i * 2, 2), 16);
  }
  return out;
}

function eqBytes(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
  return true;
}

const GOLDEN_DIR = join(__dirname, 'golden', 'aead');

describe('AES-256-GCM - golden vectors', () => {
  const files = readdirSync(GOLDEN_DIR).filter((f) => f.endsWith('.json'));
  expect(files.length).toBeGreaterThanOrEqual(2);

  for (const f of files) {
    it(`golden: ${f}`, async () => {
      const golden = JSON.parse(
        readFileSync(join(GOLDEN_DIR, f), 'utf-8'),
      ) as Record<string, string>;

      // key
      const key =
        typeof golden.key_b64 === 'string'
          ? b64(golden.key_b64)
          : typeof golden.key_hex === 'string'
            ? hex(golden.key_hex)
            : typeof golden.wrap_key_b64 === 'string'
              ? b64(golden.wrap_key_b64)
              : (() => {
                  throw new Error('missing key field');
                })();

      // nonce
      const nonce =
        typeof golden.nonce_b64 === 'string'
          ? b64(golden.nonce_b64)
          : typeof golden.nonce_hex === 'string'
            ? hex(golden.nonce_hex)
            : typeof golden.wrap_nonce_b64 === 'string'
              ? b64(golden.wrap_nonce_b64)
              : (() => {
                  throw new Error('missing nonce field');
                })();

      // plaintext
      const plaintext =
        typeof golden.plaintext_b64 === 'string'
          ? b64(golden.plaintext_b64)
          : typeof golden.master_key_b64 === 'string'
            ? b64(golden.master_key_b64)
            : (() => {
                throw new Error('missing plaintext field');
              })();

      // aad（可选）
      const aad =
        typeof golden.aad_b64 === 'string' ? b64(golden.aad_b64) : new Uint8Array(0);

      // 加密：可能用 wrapped_key（ciphertext||tag 拼接）或拆开的 ciphertext+tag
      const enc = await aesGcmEncrypt(key, nonce, plaintext, aad);

      if (typeof golden.expected_wrapped_key_b64 === 'string') {
        const expectedWrapped = b64(golden.expected_wrapped_key_b64);
        const got = new Uint8Array(enc.ciphertext.length + enc.tag.length);
        got.set(enc.ciphertext, 0);
        got.set(enc.tag, enc.ciphertext.length);
        expect(got.length).toBe(expectedWrapped.length);
        expect(eqBytes(got, expectedWrapped)).toBe(true);
      } else {
        const expectedCt = b64(golden.expected_ciphertext_b64);
        const expectedTag = b64(golden.expected_tag_b64);
        expect(eqBytes(enc.ciphertext, expectedCt)).toBe(true);
        expect(eqBytes(enc.tag, expectedTag)).toBe(true);
      }

      // 反向：解密回明文
      const decrypted = await aesGcmDecrypt(key, nonce, enc.ciphertext, enc.tag, aad);
      expect(eqBytes(decrypted, plaintext)).toBe(true);
    });
  }
});

describe('AES-256-GCM - tamper detection', () => {
  it('tampered tag → decrypt throws', async () => {
    const key = new Uint8Array(32).fill(0x01);
    const nonce = new Uint8Array(12).fill(0x02);
    const pt = new TextEncoder().encode('hello aun');
    const aad = new TextEncoder().encode('aad');
    const { ciphertext, tag } = await aesGcmEncrypt(key, nonce, pt, aad);
    const badTag = new Uint8Array(tag);
    badTag[0] ^= 0xff;
    await expect(aesGcmDecrypt(key, nonce, ciphertext, badTag, aad)).rejects.toThrow();
  });

  it('tampered aad → decrypt throws', async () => {
    const key = new Uint8Array(32).fill(0x01);
    const nonce = new Uint8Array(12).fill(0x02);
    const pt = new TextEncoder().encode('hello aun');
    const aad = new TextEncoder().encode('aad');
    const { ciphertext, tag } = await aesGcmEncrypt(key, nonce, pt, aad);
    const wrongAad = new TextEncoder().encode('AAD');
    await expect(aesGcmDecrypt(key, nonce, ciphertext, tag, wrongAad)).rejects.toThrow();
  });
});
